### PR TITLE
Runtime reducer estimator fixes

### PR DESCRIPTION
**Pull request by sid-kap**
Some changes that I made when testing the reducer estimator

- Instead of 4 runtime reducer estimators, now there is only one HRavenRuntimeReducerEstimator with sane defaults (prediction is input-scaled, uses median task time per job as goal). 
- Use Double arithmetic instead of Long.
- Update HRaven version so that I can get Task history. (HRaven versioning is weird, but this seems to be compatible with the version of HRaven in source.)
- Improve logging.

### CODE REVIEW

File: scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala [hunk 1]
```diff
@@ -43,69 +72,135 @@ trait RuntimeEstimationScheme {
...
     if (inputBytes == 0) {
+      LOG.warn("Input bytes is zero in current step.")
       None
     } else {
       // numReducers = time-per-byte * numBytes / desiredRuntime
-      typicalTimeToByteRatio.map { t => (t.toDouble * inputBytes / desiredRuntime).ceil.toInt }
+      val estimate = typicalTimeToByteRatio.map { t => (t.toDouble * inputBytes / desiredRuntime).ceil.toInt }
+
+      LOG.info(s"""
+        | - History: $history
```
Comment (sid-kap) [inline scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala]: This is too much debugging. Will get rid of "$history".

File: version.sbt [hunk 2]
```diff
@@ -1 +1 @@
-version in ThisBuild := "0.15.1-SNAPSHOT"
+version in ThisBuild := "0.15.8-SNAPSHOT"
```
Comment (sid-kap) [inline version.sbt]: I will revert this

File: scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala [hunk 3]
```diff
@@ -8,11 +8,40 @@ import org.apache.hadoop.mapred.JobConf
...
+
+  /**
+   * Whether to ignore the input size of the data.
+   * If true, RuntimeReducerEstimator uses a non-input scaled estimator.
+   * If false, RuntimeReducerEstimator uses an input-scaled estimator
+   * first, and uses a non-input-scaled estimator as a fallback.
+   * Default is false.
+   */
+  def getRuntimeIgnoreInputSize(conf: JobConf): Boolean = {
+    val RuntimeIgnoreInputSize = "scalding.reducer.estimator.runtime.ignore.input.size"
```
Comment (sid-kap) [inline scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala]: Will make this less verbose by changing it to "IgnoreInputSize"

File: scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala [hunk 4]
```diff
@@ -8,11 +8,40 @@ import org.apache.hadoop.mapred.JobConf
 import org.slf4j.LoggerFactory
 
 object RuntimeReducerEstimator {
-  val RuntimePerReducer = "scalding.reducer.estimator.runtime.per.reducer"
-  val defaultRuntimePerReducer = 10 * 60 * 1000 // 10 mins
```
Comment (rubanm) [inline scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala]: Any reason for removing these? Might be useful for outside code to fetch the defaults.

File: scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala [hunk 5]
```diff
@@ -43,69 +72,134 @@ trait RuntimeEstimationScheme {
...
 trait BasicRuntimeReducerEstimator extends HistoryReducerEstimator {
   import RuntimeReducerEstimator._
 
+  private val LOG = LoggerFactory.getLogger(this.getClass)
+
   def runtimeEstimationScheme: RuntimeEstimationScheme
 
   def estimateReducers(info: FlowStrategyInfo, history: Seq[FlowStepHistory]): Option[Int] = {
-    val reduceTimes: Seq[Seq[Long]] = getReduceTimes(history)
+    val reduceTimes: Seq[Seq[Double]] = getReduceTimes(history).map(xs => xs.map(_.toDouble))
```
Comment (rubanm) [inline scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala]: Should the `toDouble` conversion be moved to `getReducerTimes`?

File: scalding-hraven/src/main/scala/com/twitter/scalding/hraven/reducer_estimation/HRavenHistoryService.scala [hunk 6]
```diff
@@ -75,18 +75,29 @@ object HRavenHistoryService extends HistoryService {
...
+    Try(client.fetchFlowsWithConfig(cluster, user, batch, signature, nFetch, RequiredJobConfigs: _*))
+      .flatMap { flows =>
+        Try {
+          // Ugly mutable code to add task info to flows
+          flows.asScala.foreach { flow =>
+            flow.getJobs.asScala.foreach { job =>
+              val tasks = client.fetchTaskDetails(flow.getCluster, job.getJobId)
+              job.addTasks(tasks)
+            }
+          }
```
Comment (rubanm) [inline scalding-hraven/src/main/scala/com/twitter/scalding/hraven/reducer_estimation/HRavenHistoryService.scala]: So looks like this was a bug affecting the existing RatioBasedEstimator as well?

File: scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala (hunk 4 above)
Comment (rubanm) [inline scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala]: Ah scratch that. This looks fine actually.

File: project/Build.scala [hunk 7]
```diff
@@ -410,7 +410,7 @@ object ScaldingBuild extends Build {
 
   lazy val scaldingHRaven = module("hraven").settings(
     libraryDependencies ++= Seq(
-      "com.twitter.hraven" % "hraven-core" % hravenVersion,
+      "com.twitter.hraven" % "hraven-core" % hravenVersion exclude("javax.jms", "jms") exclude("com.sun.jdmk", "jmxtools") exclude("com.sun.jmx", "jmxri"),
```
Comment (rubanm) [inline project/Build.scala]: nit - excludes to be in separate lines

File: scalding-hraven/src/main/scala/com/twitter/scalding/hraven/reducer_estimation/HRavenHistoryService.scala [hunk 8]
```diff
@@ -75,18 +75,29 @@ object HRavenHistoryService extends HistoryService {
...
+
+          val successfulFlows = flows.asScala.filter(_.getHdfsBytesRead > 0).take(max)
+          if (successfulFlows.isEmpty) {
+            LOG.warn("Unable to find any successful flows in the last " + nFetch + " jobs.")
+          }
+          successfulFlows
+        }
+      } recoverWith {
```
Comment (rubanm) [inline scalding-hraven/src/main/scala/com/twitter/scalding/hraven/reducer_estimation/HRavenHistoryService.scala]: recoverWith here should still apply to the outer Try right?

Comment (ianoc) [inline scalding-hraven/src/main/scala/com/twitter/scalding/hraven/reducer_estimation/HRavenHistoryService.scala]: Also always use .'s between methods. (I see the old didn't have it, but it should have had)

**Comment by rubanm** [comment]
Thanks for the fixes. Would be good to have some basic unit tests for RuntimeReducerEstimator, similar to the ones added for RatioBasedEstimator.

File: scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala [hunk 9]
```diff
@@ -8,11 +8,40 @@ import org.apache.hadoop.mapred.JobConf
...
+   * Whether to use the median or the mean in the runtime estimation process.
+   * Default is median.
+   */
+  def getRuntimeEstimationScheme(conf: JobConf): RuntimeEstimationScheme = {
+    val key = "scalding.reducer.estimator.runtime.estimation.scheme"
+    val default = "median"
+    conf.get(key, default) match {
+      case "mean" => MeanEstimationScheme
+      case "median" => MedianEstimationScheme
+      case _ => MedianEstimationScheme
```
Comment (rubanm) [inline scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala]: Should we just throw here? I feel it's better for users to be precise if they are providing the scheme than us silently picking median.

File: scalding-hraven/src/main/scala/com/twitter/scalding/hraven/reducer_estimation/HRavenHistoryService.scala (hunk 6 above)
Comment (sid-kap) [inline scalding-hraven/src/main/scala/com/twitter/scalding/hraven/reducer_estimation/HRavenHistoryService.scala]: I don't think the ratio-based estimator uses task-level data

File: scalding-hraven/src/main/scala/com/twitter/scalding/hraven/reducer_estimation/HRavenHistoryService.scala (hunk 8 above)
Comment (sid-kap) [inline scalding-hraven/src/main/scala/com/twitter/scalding/hraven/reducer_estimation/HRavenHistoryService.scala]: Sure, I'll add the dot.
recoverWith still applies to the outer Try, I just indented incorrectly so it looks like it changed. I'll fix the indentation.

File: scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala (hunk 9 above)
Comment (sid-kap) [inline scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala]: Sure, that makes sense

File: scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala (hunk 5 above)
Comment (sid-kap) [inline scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala]: Sure, might as well since it's common to both places where getReduceTimes is used

**Comment by sid-kap** [comment]
Should I add unit tests in this PR or make a separate PR?

File: scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala (hunk 4 above)
Comment (sid-kap) [inline scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala]: This will be useful for writing unit tests, so I think I'll move it out of the method.

**Comment by rubanm** [comment]
@sid-kap ideally this PR I'd say so we can exercise the tests as part of CI.

File: scalding-hadoop-test/src/test/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimatorTest.scala [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+  def fetchHistory(info: FlowStrategyInfo, maxHistory: Int): Try[Seq[FlowStepHistory]] =
+    Success(
+      Seq(
+        makeHistory(inputSize * 2, 0, List(10, 1000, 3000)),
+        makeHistory(inputSize / 2, 0, List(10, 200, 400)),
+        makeHistory(inputSize * 4, 0, List(10, 2400, 3000))))
+}
+
+class Estimator1 extends RuntimeReducerEstimator {
+  override val historyService = HistoryService1
```
Comment (rubanm) [inline scalding-hadoop-test/src/test/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimatorTest.scala]: Probably good to also add test cases for 'empty history' and 'failed to fetch history', given we not share anything with RatioBasedEstimator so don't have coverage for that here.

File: scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala [hunk 11]
```diff
@@ -43,69 +74,136 @@ trait RuntimeEstimationScheme {
...
+    val typicalJobTime: Option[Double] = runtimeEstimationScheme.estimateJobTime(jobTimes.flatten)
 
-    val desiredRuntime = getRuntimePerReducer(info.step.getConfig)
-    typicalJobTime.map { t => (t.toDouble / desiredRuntime).ceil.toInt }
+    val desiredRuntime: Long = getRuntimePerReducer(info.step.getConfig)
+
+    val estimate = typicalJobTime.map { t: Double => (t / desiredRuntime).ceil.toInt }
+
+    LOG.info(s"""
+      | - Job times: $jobTimes
```
Comment (rubanm) [inline scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala]: Just wondering if logging the whole sequence might be too verbose? The other values below are `Option`s so should be fine.

**Comment by rubanm** [comment]
Some minor comments. This lgtm to merge after that. Thanks for adding the tests.

File: scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala (hunk 11 above)
Comment (sid-kap) [inline scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala]: Sure, I can remove this

File: scalding-hadoop-test/src/test/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimatorTest.scala (hunk 10 above)
Comment (sid-kap) [inline scalding-hadoop-test/src/test/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimatorTest.scala]: Fixed in latest commit

File: project/Build.scala [hunk 12]
```diff
@@ -410,7 +410,21 @@ object ScaldingBuild extends Build {
...
   lazy val scaldingHRaven = module("hraven").settings(
     libraryDependencies ++= Seq(
-      "com.twitter.hraven" % "hraven-core" % hravenVersion,
+      "com.twitter.hraven" % "hraven-core" % hravenVersion
+        // These transitive dependencies cause sbt to give a ResolveException
+        // because they're not available on Maven. We don't need them anyway.
+        // See https://github.com/twitter/cassie/issues/13
+        exclude("javax.jms", "jms")
+        exclude("com.sun.jdmk", "jmxtools")
+        exclude("com.sun.jmx", "jmxri")
```
Comment (rubanm) [inline project/Build.scala]: I think most of these excludes would not have been needed if there was a separate `hraven-client` module. There is none sadly.
